integration/pause_test.go

```diff
@@ -0,0 +1,259 @@
+			Name: addonsv1alpha1.DefaultAddonOperatorName,
+		}, addonOperator)
+		if err != nil {
+			return err
+		}
+
+		// update field
+		addonOperator.Spec.Pause = true
+
+		err = integration.Client.Update(ctx, addonOperator)
+		if err != nil {
+			return err
+		}
+		return nil
+	})
+	require.NoError(t, err)
+
+	// Wait for AddonOperator to report Unavailable condition
+	err = integration.Client.Get(ctx, client.ObjectKey{
+		Name: addonsv1alpha1.DefaultAddonOperatorName,
+	}, addonOperator)
+	require.NoError(t, err)
+	err = integration.WaitForObject(
+		t, 2*time.Minute, addonOperator, "to be Unavailable",
+		func(obj client.Object) (done bool, err error) {
+			a := obj.(*addonsv1alpha1.AddonOperator)
+			return testutil.AssertReasonForConditionUnavailable(
+				a.Status.Conditions,
+				controllers.AddonOperatorPaused,
+			), nil
+		})
+	require.NoError(t, err)
+
+	// Load Addon
+	uuid := "9c4a3192-6a79-4782-93dd-636e4d308852"
+	addonName := fmt.Sprintf("addon-%s", uuid)
+	addonNamespace := fmt.Sprintf("namespace-%s", uuid)
+
+	addon := &addonsv1alpha1.Addon{
+		ObjectMeta: v1.ObjectMeta{
+			Name: addonName,
+		},
+		Spec: addonsv1alpha1.AddonSpec{
+			DisplayName: addonName,
+			Namespaces: []addonsv1alpha1.AddonNamespace{
+				{Name: addonNamespace},
+			},
+			Install: addonsv1alpha1.AddonInstallSpec{
+				Type: addonsv1alpha1.OLMOwnNamespace,
+				OLMOwnNamespace: &addonsv1alpha1.AddonInstallOLMOwnNamespace{
+					AddonInstallOLMCommon: addonsv1alpha1.AddonInstallOLMCommon{
+						Namespace:          addonNamespace,
+						CatalogSourceImage: referenceAddonCatalogSourceImageWorking,
+						PackageName:        "reference-addon",
+						Channel:            "alpha",
+					},
+				},
+			},
+		},
+	}
+
+	// Create Addon
+	err = integration.Client.Create(ctx, addon)
+	require.NoError(t, err)
+
+	t.Cleanup(func() {
+		err := integration.Client.Delete(ctx, addon, client.PropagationPolicy("Foreground"))
+		if client.IgnoreNotFound(err) != nil {
+			t.Logf("could not clean up Addon %s: %v", addon.Name, err)
+		}
+
+		// Resume reconciliation by setting `pause: false`
+		err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+			addonOperator = &addonsv1alpha1.AddonOperator{}
+			err := integration.Client.Get(ctx, client.ObjectKey{
+				Name: addonsv1alpha1.DefaultAddonOperatorName,
+			}, addonOperator)
+			if err != nil {
+				return err
+			}
+
+			// update field
+			addonOperator.Spec.Pause = false
+
+			err = integration.Client.Update(ctx, addonOperator)
+			if err != nil {
+				return err
+			}
+			return nil
+		})
+		if err != nil {
+			t.Logf("failed to set `pause: false`: %v", err)
+		}
+	})
+
+	// Addon expected to be stuck in Pending phase
+	err = integration.WaitForObject(
+		t, 2*time.Minute, addon, "to be Available",
+		func(obj client.Object) (done bool, err error) {
+			a := obj.(*addonsv1alpha1.Addon)
+			return meta.IsStatusConditionTrue(
+				a.Status.Conditions, addonsv1alpha1.Available), nil
+		})
+	require.Error(t, err)
```

Comment:
The problem here is that when an `Addon` is created after `paused: true` is set on `AddonOperator`, its just going to be stuck in `Pending` (which is expected behaviour, because we don't want to see reconciliation happening). The `Addon` reports a Paused condition only when `paused: true` is set on the given `Addon` itself, and not when its paused because the `AddonOperator` is set to pause.

The reason why I added a timer of 2mins was because even when there's no pause, the Addon is in Pending state for a while until the dependent resources are created. The idea here was to check if "Addon is stuck _actually_ in Pending".

Reply:
I would argue, that if you can't be sure that the Addon is actually paused, we provide a bad interface.
While the Addon object needs to stop modifying resources during reconciliation, we should still update the Addon's status to inform clients what is happening.